lib/membrane_mp4/muxer/cmaf.ex

```diff
@@ -30,138 +38,224 @@ defmodule Membrane.MP4.Muxer.CMAF do
-    caps = %Membrane.CMAF.Track{
-      content_type:
-        case caps.content do
-          %AVC1{} -> :video
-          %AAC{} -> :audio
-        end,
-      header:
-        caps
-        |> Map.take([:timescale, :width, :height, :content])
-        |> Header.serialize()
-    }
+  def handle_end_of_stream(Pad.ref(:input, _track_id) = pad, ctx, state) do
+    sample = state.pad_to_track_data[pad].buffer_awaiting_duration
 
-    # returning new output caps action should be postponed so that
-    # the discontinuity event the action will trigger will come right after the last buffer representing old
-    # input caps is sent, therefore if there are cached samples we need to cache both caps:
-    # - old input caps - must be used to generate a segment from the cached samples, have to be cached as the next handle_process call will use new input caps
-    # - new output caps - must follow freshly generated segment from the old input caps
-    {caps, state} =
-      cond do
-        # cache caps only if there are cached samples
-        caps != ctx.pads.output.caps and state.samples != [] ->
-          {[], %{state | old_input_caps: ctx.pads.input.caps, new_output_caps: caps}}
+    sample_metadata =
+      Map.put(sample.metadata, :duration, hd(state.samples[pad]).metadata.duration)
 
-        caps != ctx.pads.output.caps ->
-          {[caps: {:output, caps}], state}
+    sample = %Buffer{sample | metadata: sample_metadata}
 
-        true ->
-          {[], state}
+    state = update_in(state, [:samples, pad], &[sample | &1])
+
+    processing_finished? =
+      ctx.pads |> Map.drop([:output, pad]) |> Map.values() |> Enum.all?(& &1.end_of_stream?)
+
+    if processing_finished? do
+      with {:ok, segment, state} <- Segment.Helper.take_all_samples(state) do
+        {buffer, state} = generate_segment(segment, ctx, state)
+        {{:ok, buffer: {:output, buffer}, end_of_stream: :output}, state}
+      else
+        {:error, :not_enough_data} -> {{:ok, end_of_stream: :output}, state}
       end
+    else
+      state = put_in(state, [:pad_to_track_data, pad, :end_timestamp], nil)
 
-    {{:ok, caps ++ [redemand: :output]}, state}
+      {{:ok, redemand: :output}, state}
+    end
   end
 
-  @impl true
-  def handle_end_of_stream(:input, ctx, state) do
-    last_dts = hd(state.samples).dts
+  defp generate_caps(state) do
+    tracks = Enum.map(state.pad_to_track_data, fn {_pad, track_data} -> track_data.track end)
 
-    {buffer, state} =
-      generate_segment(ctx.pads.input.caps, %{dts: last_dts, metadata: %{}}, state)
+    header = Header.serialize(tracks)
 
-    {{:ok, buffer: {:output, buffer}, end_of_stream: :output}, state}
+    content_type =
+      tracks
+      |> Enum.map(fn
+        %{content: %AAC{}} -> :audio
+        %{content: %AVC1{}} -> :video
+      end)
+      |> then(fn
+        [item] -> item
+        list -> list
+      end)
+
+    %Membrane.CMAF.Track{
+      content_type: content_type,
+      header: header
+    }
```

Comment:
We should release new `membrane_cmaf_format` and update `mix.exs`, as it's now incorrect given the spec from `0.4.0`.

Reply:
I'm fully aware of all that, but I need an approval on the PR to `membrane_cmaf_format`